// libjs/lib/platforms/linux/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// # x86_64 Linux
const SYS_1 = require("./SYS");
exports.SYS = SYS_1.default;
const MCAST_1 = require("../../consts/MCAST");
exports.MCAST = MCAST_1.default;
const basetypes_1 = require("./basetypes");
exports.NULL = basetypes_1.NULL;
exports.int8 = basetypes_1.int8;
exports.uint8 = basetypes_1.uint8;
exports.int16 = basetypes_1.int16;
exports.uint16 = basetypes_1.uint16;
exports.int32 = basetypes_1.int32;
exports.uint32 = basetypes_1.uint32;
exports.int64 = basetypes_1.int64;
exports.uint64 = basetypes_1.uint64;
exports.size_t = basetypes_1.size_t;
exports.optval_t = basetypes_1.optval_t;
exports.pid_t = basetypes_1.pid_t;
exports.pointer_t = basetypes_1.pointer_t;
exports.time_t = basetypes_1.time_t;
const ipv4_1 = require("../../structs/ipv4");
exports.ipv4 = ipv4_1.default;
const stat_1 = require("../../structs/stat");
exports.statStruct = stat_1.default;
const in_addr_1 = require("../../structs/in_addr");
exports.in_addr = in_addr_1.default;
const sockaddr_in_1 = require("../../structs/sockaddr_in");
exports.sockaddr_in = sockaddr_in_1.default;
const sockaddr_1 = require("../../structs/sockaddr");
exports.sockaddr = sockaddr_1.default;
const epoll_event_1 = require("../../structs/epoll_event");
exports.epoll_event = epoll_event_1.default;
const ipc_perm_1 = require("../../structs/ipc_perm");
exports.ipc_perm = ipc_perm_1.default;
const shmid_ds_1 = require("../../structs/shmid_ds");
exports.shmid_ds = shmid_ds_1.default;
const utimbuf_1 = require("../../structs/utimbuf");
exports.utimbuf = utimbuf_1.default;
const timeval_1 = require("../../structs/timeval");
exports.timeval = timeval_1.default;
const timevalarr_1 = require("../../structs/timevalarr");
exports.timevalarr = timevalarr_1.default;
const timespec_1 = require("../../structs/timespec");
exports.timespec = timespec_1.default;
const linux_dirent64_1 = require("../../structs/linux_dirent64");
exports.linux_dirent64 = linux_dirent64_1.default;
const inotify_event_1 = require("../../structs/inotify_event");
exports.inotify_event = inotify_event_1.default;
// Maximal file path string length;
exports.PATH_MAX = 4096;
// Is little-endian.
exports.isLE = true;
const timespecarr = timevalarr_1.default;
exports.timespecarr = timespecarr;
