// libjs/lib/process.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// Process
const platform_1 = require("./platform");
const libsys_1 = require("./libsys");
const createSyscall = (num) => () => libsys_1.libsys.syscall(num);
const createAsyscall = (num) => (callback) => libsys_1.libsys.asyscall(num, callback);
// ### getpid
//
//     getpid(): number
//
// Get process ID.
//
exports.getpid = createSyscall(platform_1.SYS.getpid);
exports.getpidAsync = createAsyscall(platform_1.SYS.getpid);
// ### getppid
//
//     getppid(): number
//
// Get parent process ID.
//
exports.getppid = createSyscall(platform_1.SYS.getppid);
exports.getppidAsync = createAsyscall(platform_1.SYS.getppid);
// ### getuid
//
//     getuid(): number
//
// Get parent user ID.
//
exports.getuid = createSyscall(platform_1.SYS.getuid);
exports.getuidAsync = createAsyscall(platform_1.SYS.getuid);
// ### geteuid
//
//     geteuid(): number
//
// Get parent real user ID.
//
exports.geteuid = createSyscall(platform_1.SYS.geteuid);
exports.geteuidAsync = createAsyscall(platform_1.SYS.geteuid);
// ### getgid
//
//     getgid(): number
//
// Get group ID.
//
exports.getgid = createSyscall(platform_1.SYS.getgid);
exports.getgidAsync = createAsyscall(platform_1.SYS.getgid);
// ### getgid
//
//     getegid(): number
//
// Get read group ID.
//
exports.getegid = createSyscall(platform_1.SYS.getegid);
exports.getegidAsync = createAsyscall(platform_1.SYS.getegid);
// ### sched_yield
exports.sched_yield = createSyscall(platform_1.SYS.sched_yield);
// ## sleep
function nanosleep(seconds, nanoseconds) {
    const buf = platform_1.timespec.pack({
        tv_sec: [seconds, 0],
        tv_nsec: [nanoseconds, 0],
    });
    return libsys_1.libsys.syscall(platform_1.SYS.nanosleep, buf, platform_1.NULL);
}
exports.nanosleep = nanosleep;
